MWPW-172706 - create a CMR for each PR to the main branch

tusc-adbe opened the pull request:
~Do not merge.  Testing functionality.~
~**Before merging:**~
* ~Update `branches` to `main`~ - DONE


**Overview**
* This workflow will create CMRs for any changes to M@S Studio or M@S IO
* The workflow runs when a PR for the main branch is opened or closed, and contains changes under io/ or studio/
* On PR open
  * The workflow determines exactly what changed and create a CMR for M@S Studio, M@S IO, or both.
    * Transaction numbers for those CMRs are written to the PR comments.  The workflow will reference these transaction numbers when closing the CMRs.
  * A notification will be sent to the Slack #merch-at-scale channel
* On PR close
  * The workflow will check PR comments to determine whether any CMRs were opened, and will update and close them
    * If the PR was merged, the CMRs will be closed as Successful; if the PR was closed without merging the CMRs will be Cancelled.
  * A notification will be sent to the Slack #merch-at-scale channel (only if the PR was merged)
* Future enhancements
  * Expand this to include other paths (this may require creating and onboarding new services in the Service Registry
  * Refactor this workflow so all actions take place on PR close.  This will reduce CMR noise and reduce the time between opening a PR and closing it.
  * These will be done in separate JIRA tickets and PRs

Commit:
MWPW-172706 - create a PR for each PR to the main branch.  First commit, for testing.

Commit:
Merge branch 'main' into MWPW-172706

Commit:
MWPW-172706 - rewrite path logic to remove dependency on external action

Commit:
Merge branch 'main' into MWPW-172706

Commit:
MWPW-172706 - remove permissions from the job step

tusc-adbe committed:
MWPW-172706 - refactor new workflow to use ES module syntax, and temp disable the other workflows for testing (#458)

* MWPW-172706-incoming-PRs - testing actions in the MWPW-172706 branch.  Change under the MAS Studio folder.

* MWPW-172706 - test workflow by making changes to both watched paths

* MWPW-172706 - refactor to use ES module syntax

* MWPW-172706 - temp remove test files

* MWPW-172706 - temporarily disable the other workflows for testing

---------

Co-authored-by: Gary Tuscany <tuscany@adobe.com>

tusc-adbe committed:
MWPW-172706 - test workflow by making changes to both watched paths (#467)

tusc-adbe committed:
MWPW-172706 - test workflow by making changes to both watched paths (#469)

Commit:
MWPW-172706 - switch from pull_request_target to pull_request

tusc-adbe committed:
MWPW-172706 - test workflow by making changes to both watched paths (#470)

Commit:
MWPW-172706 - update servicenow stage URLs

Commit:
MWPW-172706 - update workflow to pass the serviceIDs correctly

Commit:
MWPW-172706 - rework syntax since github didn't like the last one.

tusc-adbe committed:
MWPW-172706 - test workflow by making changes to both watched paths (#472)

Commit:
MWPW-172706 - only run detect-changes step on PR open

tusc-adbe committed:
MWPW-172706 - test workflow by making changes to both watched paths (#473)

Commit:
MWPW-172706 - switch to temp service IDs for testing

Commit:
MWPW-172706 - another update to switch to temp service IDs for testing

tusc-adbe committed:
MWPW-172706 - test workflow by making changes to both watched paths (#476)

Commit:
MWPW-172706 - add ENVIRONMENT variable to easily assign correct IMS and SNOW URLs, and update Slack message accordingly

Commit:
MWPW-172706 - update the slack webhook variable to be more generic

tusc-adbe committed:
MWPW-172706 - test workflow by making changes to both watched paths (#479)

Commit:
MWPW-172706 - update the CMR test results message

Commit:
MWPW-172706 - swtich back to test Slack channel for notifications

tusc-adbe committed:
MWPW-172706 - test workflow by making changes to only one watched paths (#483)

Commit:
Merge branch 'MWPW-172706' of https://github.com/adobecom/mas into MWPW-172706

tusc-adbe committed:
MWPW-172706 - test workflow by making changes to only one watched paths, io (#484)

tusc-adbe committed:
MWPW-172706 - test workflow by making changes to an unwatched path (#485)

Commit:
MWPW-172706 - switch to prod values for service registry IDs, IMS and ServiceNow

Commit:
MWPW-172706 - switch to prod secrets

Commit:
Merge branch 'main' into MWPW-172706

Commit:
MWPW-172706 - switch to back to stage secrets for additional stage testing

Commit:
MWPW-172706 - major refactor.  Create 1 or 2 separate CMRs depending on the updated paths, instead of one CMR with one or two services

tusc-adbe committed:
MWPW-172706 - test prod CMR with one watched path, io (#489)

Commit:
MWPW-172706 - more refactoring, for PR closure

tusc-adbe committed:
MWPW-172706 - test prod CMR with one watched path, io (#490)

Commit:
MWPW-172706 - more refactoring, for Slack notifications

tusc-adbe committed:
MWPW-172706 - test prod CMR with one watched path, studio (#491)

tusc-adbe committed:
MWPW-172706 - test prod CMR with both watched paths (#492)

Commit:
Merge branch 'main' into MWPW-172706

Commit:
MWPW-172706 - temp disable the other workflows for testing

tusc-adbe committed:
MWPW-172706 - prep for prod testing

tusc-adbe committed:
MWPW-172706 - more prep for prod testing, and temp disable another workflow

tusc-adbe committed:
MWPW-172706 - test prod CMR with both watched paths (#606)

tusc-adbe committed:
MWPW-172706 - more tweaks for prod testing

tusc-adbe committed:
MWPW-172706 - test prod CMR with both watched paths (#607)

tusc-adbe committed:
MWPW-172706 - some cleanup of old test values

tusc-adbe committed:
MWPW-172706 - consolidate duplicated code into a matrix strategy

tusc-adbe committed:
MWPW-172706 - switch back to stage for more testing

tusc-adbe committed:
Merge branch 'main' into MWPW-172706

tusc-adbe committed:
MWPW-172706 - merge from main, temp diabled more new workflows while testing

tusc-adbe committed:
MWPW-172706 - matrix logic tweaks, tweak some syntax

tusc-adbe committed:
MWPW-172706 - test new matrix logic, stage CMR with both watched paths (#692)

tusc-adbe committed:
MWPW-172706 - test new matrix logic, IO path only (#693)

tusc-adbe committed:
MWPW-172706 - test new matrix logic, Studio path only (#696)

tusc-adbe committed:
MWPW-172706 - test new matrix logic, neither IO nor Studio path (#697)

tusc-adbe committed:
Merge branch 'main' into MWPW-172706

3ch023 commented on .github/workflows/servicenow.yaml:
```diff
@@ -0,0 +1,185 @@
+          echo "Changed files:"
+          echo "$CHANGED_FILES"
+          
+          # Check if any files match the io/** pattern
```
can we add here /io OR /web-components?

tusc-adbe replied:
Yes and no.  :)  This workflow can be expanded to additional paths but unfortunately it's not as simple as dropping in a new path here.  There's more logic below that needs to be updated, and if the new path relates to a separate service in the Service Registry, then that service needs to be created and onboarded first.

npeltier replied:
yes, i guess we will need new service registrations for web-component & probably ost too

3ch023 commented on .github/workflows/servicenow.yaml:
```diff
@@ -0,0 +1,185 @@
+  pull_request:
+    types:
+      - opened
+      - closed
```
shouldn't it trigger on merge?

tusc-adbe replied:
No.  The CMR is opened when the PR is opened.  If we only update and close the CMR on `merge` then we miss the case where a PR is closed without merging, and we'll end up with a lot of orphaned CMRs.  There's logic later in the workflow to determine whether or not it was actually merged.

tusc-adbe replied:
No.  CMRs will be created when the PR is opened.  When a PR is closed, we need to close those CMRs.  If we use `merge` here, then we'll miss closing CMRs for PRs that are closed without merging, but `close` will catch both merged PRs and PRs that were closed without merging.  There's logic later in the workflow to determine whether the PR was merged: if it was merged then we close the CMR as successful, if it was closed without merging then we cancel the CMR.

3ch023 commented on .github/workflows/servicenow.yaml:
```diff
@@ -0,0 +1,185 @@
+      - opened
+      - closed
+    branches:
+      - MWPW-172706
```
i guess it will be updated later :)

tusc-adbe replied:
Right, updating this to `main` is the *absolute* last thing to update before merging this PR.

3ch023 reviewed (commented):
Hi @tusc-adbe , I checked it a bit and left some minor comments.
could you cleanup PR and make it ready for merging? Nico can give it a look and we merge afterwards

tusc-adbe committed:
MWPW-172706 - revert temp changes for testing

tusc-adbe committed:
Merge branch 'main' into MWPW-172706

tusc-adbe committed:
MWPW-172706 - switch environment-specific variable from stage to prod

tusc-adbe committed:
MWPW-172706 - code update needed, manually reverse 8a08b0c and 80cdb09 for further testing

tusc-adbe committed:
MWPW-172706 - add logic to close CMRs if the PR is closed without merging

tusc-adbe committed:
MWPW-172706 - test normal flow, IO only (#839)

tusc-adbe committed:
MWPW-172706 - test normal flow, both paths (#840)

tusc-adbe committed:
MWPW-172706 - test normal flow, neither path, will not run the workflow (#841)

tusc-adbe committed:
MWPW-172706 - prep for final merge, revert temp changes for testing

tusc-adbe committed:
MWPW-172706 - add overview comment block

tusc-adbe commented:
Thanks @3ch023 !

I replied to your questions, added an overview in both the PR and in the main yaml file, and cleaned up all the temp stuff so it's almost ready for a final merge.

The last thing to do before merging is to update `branches` to `main` in servicenow.yaml.

cc: @npeltier

npeltier committed:
Merge branch 'main' into MWPW-172706

npeltier commented on .github/workflows/servicenow.py:
```diff
@@ -0,0 +1,328 @@
+
+    print("Closing CMR in ServiceNow...")
+
+    pr_merged_str = os.environ.get('PR_MERGED', 'true')
```
why is default value true?

tusc-adbe replied:
Just to give it a default.  :) I suppose it could be removed if you're concerned about it, but per the code the variable will always be defined when the workflow executes.

npeltier replied:
fine by me!

npeltier reviewed (approved):
lgtm

tusc-adbe committed:
Merge branch 'main' into MWPW-172706

tusc-adbe committed:
Merge branch 'main' into MWPW-172706

tusc-adbe committed:
MWPW-172706 - final update before merge

tusc-adbe committed:
MWPW-172706 - after running prettier

npeltier committed:
Merge branch 'main' into MWPW-172706

afmicka committed:
Merge branch 'main' into MWPW-172706
